server-common/src/main/java/org/apache/gravitino/server/authorization/MetadataIdConverter.java

```diff
@@ -17,15 +17,151 @@
+          MetadataObject.Type.TOPIC, Capability.Scope.TOPIC,
+          MetadataObject.Type.COLUMN, Capability.Scope.COLUMN);
+  // Maps entity type to entity class.
+  private static final Map<Entity.EntityType, Class<?>> ENTITY_CLASS_MAPPING =
+      ImmutableMap.<Entity.EntityType, Class<?>>builder()
+          .put(Entity.EntityType.METALAKE, BaseMetalake.class)
+          .put(Entity.EntityType.CATALOG, CatalogEntity.class)
+          .put(Entity.EntityType.SCHEMA, SchemaEntity.class)
+          .put(Entity.EntityType.TABLE, TableEntity.class)
+          .put(Entity.EntityType.FILESET, FilesetEntity.class)
+          .put(Entity.EntityType.MODEL, ModelEntity.class)
+          .put(Entity.EntityType.TOPIC, TopicEntity.class)
+          .put(Entity.EntityType.TAG, TagEntity.class)
+          .put(Entity.EntityType.MODEL_VERSION, ModelVersionEntity.class)
+          .put(Entity.EntityType.COLUMN, ColumnEntity.class)
+          .put(Entity.EntityType.USER, UserEntity.class)
+          .put(Entity.EntityType.GROUP, Entity.class)
+          .put(Entity.EntityType.ROLE, RoleEntity.class)
+          .build();
 
   private MetadataIdConverter() {}
 
-  public static Long doConvert(MetadataObject metadataObject) {
-    // TODO depend on MetaCache
-    return 0L;
+  /**
+   * Converts the given metadata object to metadata id.
+   *
+   * @param metadataObject The metadata object to convert.
+   * @param metalake The metalake name.
+   * @return The metadata id.
+   * @throws IOException if an error occurs while loading the entity.
+   */
+  public static Long getID(MetadataObject metadataObject, String metalake) throws IOException {
+    CatalogManager catalogManager = GravitinoEnv.getInstance().catalogManager();
+    EntityStore entityStore = GravitinoEnv.getInstance().entityStore();
+
+    return getID(metadataObject, metalake, catalogManager, entityStore);
+  }
+
+  @VisibleForTesting
+  static Long getID(
+      MetadataObject metadataObject,
+      String metalake,
+      CatalogManager catalogManager,
+      EntityStore entityStore)
+      throws IOException {
+    Preconditions.checkArgument(metadataObject != null, "Metadata object cannot be null");
+
+    MetadataObject.Type metadataType = metadataObject.type();
+    NameIdentifier ident =
+        (metadataType != MetadataObject.Type.METALAKE)
+            ? NameIdentifier.of(DOT_PATTERN.split(metalake + "." + metadataObject.fullName()))
+            : NameIdentifier.of(metadataObject.fullName());
+
+    NameIdentifier normalizedIdent =
+        normalizeCaseSensitive(ident, METADATA_SCOPE_MAPPING.get(metadataType), catalogManager);
+
+    Entity.EntityType entityType = getEntityType(metadataType);
+
+    Entity entity;
+    try {
+      entity = entityStore.get(normalizedIdent, entityType, getEntityClass(entityType));
+    } catch (IOException e) {
+      throw new IOException("Failed to convert metadata object: " + metadataObject.fullName(), e);
```

Comment:
The message should be `failed to load entity from `entitystore`....` or something similar.